### PR TITLE
test(sync): Test syncing nodes with massive state

### PR DESCRIPTION
**Pull request by mfornet**

### CODE REVIEW

**Comment by mfornet** [comment]
This test is not passing due to very slow state part computation.

Right now, when observer2 joins the network and start syncing state it request 191 StatePart to node0. Each part take ~300ms on average on my computer to be computed for this state. Each part is computed asynchronously (however they are not being computed in parallel, they are still competing for resources)  so it takes ~57s = (191 * 300ms)  to compute all parts. 

observer2, after 10s without receiving each part start asking for all not received part again, creating a huge bottleneck for node0, which indeed is recalculating same parts again. 

Worth noticing that even if networking and sync computation are running asynchronously (and in parallel, since they can run in different threads), what I'm seeing from node0 logs is:

- Receiving 14 state parts request  (<1s)
- Computing 14 state parts (~6s)
- Send 14 state parts (<1s)
- Receiving ~180 state parts request  (<1s)
- Computing ~180 state parts (~50s)
- Receiving more parts (Not sending computed parts yet) since after 10s observer2 start requesting parts again.

After increasing waiting time to 1000s for observer2 before requesting state parts again, observer2 did manage to sync state in ~30s and bps return to "normality" (~1.2bps) but RPC for node0 remained not responsive.

**Comment by bowenwang1996** [comment]
@mfornet while network and state computation are running asynchronously, they cannot be truly parallel because both rely on database access and I believe network code actually writes into database, so they are competing for resources. According to https://github.com/iotaledger/iri/issues/1006, rocksdb support reading and writing in parallel but I am not sure that we have the right configuration.

**Comment by mfornet** [comment]
> @mfornet while network and state computation are running asynchronously, they cannot be truly parallel because both rely on database access and I believe network code actually writes into database, so they are competing for resources. According to [iotaledger/iri#1006](https://github.com/iotaledger/iri/issues/1006), rocksdb supports reading and writing in parallel but I am not sure that we have the right configuration.

One possible solution for this is scheduling changes to the database without blocking for network (it still require some thinking to check nothing get broken). 

However the main issue I wanted to raise with this comment is that state sync is a huge bottleneck, and current strategy of re-requesting state is making things worst, waiting time before re-requesting should be calculated more precisely.

Also it is possible that we can speed up the full state sync process if instead of asking for single parts we ask for range of parts, even if they are sent later as separate parts, to try to compute parts more efficiently decreasing IO access whenever possible.

**Comment by bowenwang1996** [comment]
@mfornet for sure recomputing the state parts for the same peer is very suboptimal. We definitely should cache the parts.

File: nightly/nightly.txt [hunk 1]
```diff
@@ -18,6 +18,7 @@ pytest --timeout=900 sanity/state_sync2.py
 pytest --timeout=900 sanity/state_sync3.py
 pytest --timeout=600 sanity/state_sync_routed.py manytx 100
 pytest --timeout=300 sanity/state_sync_late.py notx
+pytest --timeout=900 sanity/state_sync_massive.py
```
Comment (bowenwang1996) [inline nightly/nightly.txt:21]: If it doesn't work yet, I think there is no point adding it to nightly.

File: pytest/tests/sanity/state_sync_massive.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
+# Survive massive state sync
+#
+# Create 3 nodes, 1 validator and 2 observers tracking the single shard 0.
```
Comment (bowenwang1996) [inline pytest/tests/sanity/state_sync_massive.py:3]: what's the point of having two observers? It seems better for the purpose of testing to only have one peer to sync from.

Comment (mfornet) [reply pytest/tests/sanity/state_sync_massive.py:3]: We are only syncing from 1, but the other is necessary for node0 to start validating, it won't start without connected peers.

File: nightly/nightly.txt (hunk 1 above)
Comment (SkidanovAlex) [reply nightly/nightly.txt]: Disagree, I think the policy should be that all the tests that should be passing in a fully functional node must be in nightly.

**Comment by SkidanovAlex** [comment]
We should also consider not sending 100 requests to the same peer at once. We should have some logic that

a) Limits to say 5 parts per peer when sending
b) Bans any peer that clearly violates (a) on receiving

WDYT?

**Comment by mfornet** [comment]
> @mfornet for sure recomputing the state parts for the same peer is very suboptimal. We definitely should cache the parts.

This is done at #2515

**Comment by mfornet** [comment]
> We should also consider not sending 100 requests to the same peer at once. We should have some logic that
> 
> a) Limits to say 5 parts per peer when sending
> b) Bans any peer that clearly violates (a) on receiving
> 
> WDYT?

LGTM, I'll implement this on top of #2515 and enable this nightly test.

File: pytest/tests/sanity/state_sync_massive.py (hunk 2 above)
Comment (bowenwang1996) [reply pytest/tests/sanity/state_sync_massive.py:3]: > We are only syncing from 1, but the other is necessary for node0 to start validating, it won't start without connected peers.

Seems like a configuration issue. We can change the config to reduce min_num_peers to 0 for node0.

**Pull request merged**